### PR TITLE
Predefined variable fix

### PR DESCRIPTION
**Pull request by Legotier**
This pull request is mainly intended as a fix for #14. Even though the problem itself is quite simple, the solution was more complicated than anticipated. I implemented an entirely new sympy parser that we can now use instead of `sympy.S`. If given a complex expression as a string, it will call the internal sympy parsing function (same as `sympy.S`), but additionally:

- It only detects predefined function names if these symbols are actually used as a function. I.e., we can now have variables named `log` or `exp`, which wasn't possible previously, because these names were immediately parsed as their corresponding functions - even if they weren't used as a function but as a symbol (note that is is possible to tell `sympy.S` to interpret these names as symbols, but then they can no longer be used as functions).
- Automatically assumes that all parsed symbols are rational and nonnegative unless specified otherwise.
EDIT: In some rare cases, this seems to make sympy much slower for some reason. But it's still correct, right? Because these assumptions existed before as well, they just weren't properly implemented.
- Allows for direct parsing from a `probably.Expr` to a `sympy.Expr`, which is much faster than first converting the `probably.Expr` to a string and then parsing that string with `sympy`.

As for some additional minor improvements that came up naturally during the implementations of the new parser:

- A lot of unnecessary conversions of objects to strings and then to `sympy` objects are removed (see f8edbcee1f9a55f2bcf368fe60178efcef591a5a and 1e272bca67e8163298a261216d5661d9bfd69a75).
- A lot of our test cases are now parameterized, mainly such that they now run on both backends and rely on randomness to a lesser extent (for the latter, see b5483c52f0c3a1903cefd2fc104bdff16cabedd5).
- In our CI workflow, the tests are immediately cancelled if one fails in order to preserve precious minutes.
- Some bugfixes.

All in all, the runtime of the `sympy` backend is improved by about (using a **very** rough benchmark I did) 10%.

### CODE REVIEW

File: prodigy/distribution/generating_function.py [hunk 1]
```diff
@@ -31,6 +32,113 @@ def _term_generator(function: sympy.Poly):
...
+
+
+def _sympy_symbol(name: Any,
+                  rational=True,
+                  nonnegative=True,
+                  **kwargs) -> sympy.Symbol:
+    """
+    Creates a rational and nonnegative sympy Symbol.
+    
+    Note that `sympy.Symbol('x') == sympy.Symbol('x', rational=True, nonnegative=True)`
```
Comment (LKlinke) [inline prodigy/distribution/generating_function.py:133]: better say "evaluates to ``false``". Also a sentence describing that you are not referring to the calling convention but to the fact that comparing two variables with the same name but different domains simply yields ``false`` helps understanding that.

**Review by LKlinke — COMMENTED** [review]
is the new assumption stuff effecting the way I compute probabilities as for them I might want to have "real" valued probabilities? or square roots etc and not only rational probabilities?

**Comment by Legotier** [comment]
> is the new assumption stuff effecting the way I compute probabilities as for them I might want to have "real" valued probabilities? or square roots etc and not only rational probabilities?

I don't think so. The line
```python
SympyPGF.poisson('x', '0.3').get_probability_of('x=3')
```
evaluates to `9*exp(-3/10)/2000`, which is irrational. But it does mean that parameters can no longer represent irrational probabilities, which might be a problem.

The reason I added this is because certain simplifications are only possible if we consider the variables to be rational, e.g., the simplification ${x^9}^\frac{1}{3}$ to $x^3$.

**Review by LKlinke — COMMENTED** [review]
This seems fine.

File: prodigy/distribution/generating_function.py [hunk 2]
```diff
@@ -31,6 +32,97 @@ def _term_generator(function: sympy.Poly):
         poly -= poly.EC() * poly.EM().as_expr()
 
 
+def _parse_to_sympy(expr: Any, real=True, **kwargs) -> sympy.Expr:
```
Comment (LKlinke) [inline prodigy/distribution/generating_function.py:35]: Maybe this should be refactored to a different place than the generating function class, as it is not part of it.

**Review by LKlinke — COMMENTED** [review]
Fine for now

**Review by LKlinke — APPROVED** [review]

**Pull request merged**